### ISSUE TITLE
with onBlur validation shouldn't isValid be false if not all required fields have been entered&validated?

### ISSUE THREAD

**Issue by samlll42-github** [bluebill1049/react-hook-form#92]
First of all - I love this project, clean, modern and efficient - great job.

I started using it but I feel like something may be missing. 
When using onBlur validation, I would expect formState.isValid to be false until all the required fields have been touched & validated. Essentially so that the submit button can be disabled until the form isValid or 'isReadyToSubmit'....

Am I missing something? Is there another way to do this?

Thanks

**Comment by bluebill1049** [comment]
thank you for your kind words I will look into this issue.

**Comment by bluebill1049** [comment]
fixed and tested with `react-hook-form@3.11.9-beta.1` 

You want to do a quick test. once you can verify it's working. I will publish :)

**Comment by bluebill1049** [comment]
published https://github.com/bluebill1049/react-hook-form/releases/tag/v3.11.9 thanks for reporting.

**Comment by samlll42-github** [comment]
Thanks for the super quick turnaround! 

I tested and it does work better. However I think there Is still something that doesn't behave like I would expect.

The current implementation will set isValid once every field have been touched (and no errors are reported), regardless on whether the fields are required or not, right?  

Wouldn't it be better to only set isValid only once every 'required' field have been touched (and no errors are reported)?

Because as it is currently we still can't use isValid to disable/enable a submit button, because it would require every field to be touched even if they are not required. What do you think? or did I miss something?

Thanks again and sorry I should have been more detailed in my initial comment.

**Comment by bluebill1049** [comment]
you are right about that. i can take a look at this again

**Comment by bluebill1049** [comment]
you want to try this version?

`react-hook-form@3.11.10-beta.1`

**Comment by samlll42-github** [comment]
Thanks! I tried but I am getting isValid set to true by default when none of the fields have been touched yet, even as most of them are required...

**Comment by bluebill1049** [comment]
do you have a codesandbox that i can play around with?

**Comment by samlll42-github** [comment]
I put this one quickly together - please confirm you can access it.

https://codesandbox.io/embed/mystifying-worker-xy8pi

When we first load isValid is true...  Shouldn't the initial value be false until all required fields have been touched/validated without errors?

**Comment by bluebill1049** [comment]
I think its fixed, check it out. more complicated than I thought would be...

**Comment by bluebill1049** [comment]
try this version: https://codesandbox.io/s/react-hook-form-async-submit-validation-pjpbu

it's a bit more complicated than I initially thought

**Comment by samlll42-github** [comment]
Looks better in initial test. will let you know if we find any other problem. Thanks for fixing this!

**Comment by bluebill1049** [comment]
no worries, here is a beta version `react-hook-form@3.11.10-beta.3`

**Comment by bluebill1049** [comment]
i believe this is fixed, reopen when you find there is still issues.

**Comment by samlll42-github** [comment]
Looks there is still a problem with isValid, when we register required field outside of the input.

https://codesandbox.io/s/goofy-shirley-cw7bh
The 'test' field is registered outside of the input component, as required. Its value is handled correctly, but isValid is never set to True...

Let me know what you think...

**Comment by bluebill1049** [comment]
will take a look again

**Comment by bluebill1049** [comment]
hey @samlll42-github 

With manual registered field: 

```
useEffect(() => {
    register(
      {
        name: "test"
      },
      {
        required: true,
        min: 1
      }
    );
  }, []);
```

react hook form won't have a reference(`ref`) to the input, so it couldn't attach an event listener to watch and valid `onChange` or `onBlur`, but only `onSubmit`. you will have to `setError` or clear it manually for manually registered input.

https://react-hook-form.com/api#setError

let me know if those make sense.

cheers
bill

**Comment by samlll42-github** [comment]
It makes sense that’s what we did for now. 

but I was thinking we do update the value of required fields with setValue when they change (per the documentation), wouldn't it  be possible for the React-hook-form to set isValid to true if all required fields are set, when any of the form values change including the externally registered required fields?

Right now it stays false. Or am I missing something?

**Comment by bluebill1049** [comment]
yea I did thinking over this as well.
but I like an API to be as what they stated as and does less magic within, therefore I didn't do much extra stuff when invoking `setValue` or `setError`, but I think what you want to use is `triggerValidation`, which you can run after setValue.

let me know your thoughts.

**Comment by poteirard** [comment]
I think it's a related question so I will ask here. I have a form in mode 'onBlur' and some defaultValues and I realize that when all the values are valid "isValid" is `false`. So I have to go to every input in order to activate the validation of each one. Is that the expected behaviour? 

Is there a way to triggerValidation to all the fields?

Thanks a lot for your work :) I'm trying to move from final-form to this library :)

**Comment by bluebill1049** [comment]
Thanks, @poteirard it's not expected behavior. which version are you using?

**Comment by poteirard** [comment]
Hey @bluebill1049 thanks also for the fast response. I created a codesandbox with what I'm doing more or less. I'm using version `3.13.1`

**Comment by bluebill1049** [comment]
@poteirard thanks for the codesandbox, that really helped :)

```
React.useEffect(() => {
    triggerValidation({
      name: "email"
    });
    triggerValidation({
      name: "email2"
    });
    triggerValidation({
      name: "phone"
    });
  }, []);
```

I think this will solve your problem for now. I am planning to support `array` for `triggerValidation`. so you won't have to write it mutiple times.

**Comment by bluebill1049** [comment]
hey @samlll42-github, next release: `setValue('name', 'value', true)` will trigger validation. The third argument is a boolean for trigger validation. hope this works for your project.

**Comment by poteirard** [comment]
thanks a lot @bluebill1049 

Don't you think it will be nice to change the implementation of `triggerValidation` so when you run it without parameters it triggers the validation of every field by default? I don't know if it's possible but if it is I would be more than happy to contribute with a PR :)

**Comment by bluebill1049** [comment]
yes please @poteirard ❤️ great idea. PR welcome

**Comment by poteirard** [comment]
```ts
useEffect(() => {
    triggerValidation([{ name: "password" }, { name: "passwordConfirm" }]);
  }, []);
```

Gives me a type error as `triggerValidation` doesn't accept an array, which is correct and, correct me if I'm wrong but I think the implementation is not done yet. So I suppose I have to call `triggerValidation` for every field one by one.

Also, I think it will be nice to have an `isValid` prop for each form field. But that's another topic :)

For now, I have a very strict deadline but I promise I will come back with some help on those issues.

**Comment by bluebill1049** [comment]
@poteirard are you running the latest version. it should support `array` now :)

**Comment by bluebill1049** [comment]
@samlll42-github `setValue` with `shouldValidate` is released, you happy to close this issue?

**Comment by bluebill1049** [comment]
going to close this issue too, feel free to re-open is still have questions.

**Comment by bowd** [comment]
Not sure if related to this but I've noticed that `isValid` on `formState` doesn't update when using external validation schema (yup in this case) and `onUpdate` mode.

Here's a codesadbox: https://codesandbox.io/embed/react-hook-form-seterror-ftvrr 

This example has two fields, each required. They start off as filled in via `defaultValue` and the form is valid. Now try to edit the form and clear the password confirmation input. The form is still valid. If you press the TEST button the form becomes invalid.

Now if you change the mode to `onChange` or `onBlur` it works as expected, clearing the input causes the form to become invalid.

**Comment by bluebill1049** [comment]
@bowd i will take a look that over the weekend

**Comment by bluebill1049** [comment]
@bowd https://github.com/bluebill1049/react-hook-form/issues/117 let's discuss there

**Comment by bluebill1049** [comment]
@poteirard FYI I have made `triggerValidaiton()` with an empty argument

https://github.com/bluebill1049/react-hook-form/releases/tag/v3.18.0 :)

**Comment by poteirard** [comment]
Thanks a lot! :dancer:

**Comment by bluebill1049** [comment]
Thanks @poteirard if you find it useful, please consider staring the repo and share among the community 😄 much apprecitaed 🙏

**Comment by ArvindNikam76** [comment]
> Thanks @poteirard if you find it useful, please consider staring the repo and share among the community  much apprecitaed

disabled={Object.keys(errors).length >0  || formState.dirtyFields.size < 5}
